### ISSUE TITLE
IDEA: Remove KODI and let users use NOOBS to install OSMC or OpenELEC

### ISSUE THREAD

**Issue by matthuisman** [recalbox/recalbox-os#429]
As per messages here:
https://www.raspberrypi.org/forums/viewtopic.php?p=815954#p815954
https://www.raspberrypi.org/forums/viewtopic.php?p=815972#p815972

Would it be a good idea to remove KODI from Recalbox and allow it to be a separate NOOBS package?

Pretty much, you guys would ship vanilla NOOBS that allows downloading & installing the other OS's.

User could then install the other OS's they require.

Recalbox could detect which OS's are installed and then have a new emulator "OS's" listed.
In this listing, the rom listings would be for the other OS's "Raspbian, KODI" etc.

Clicking on the OS, would simply reboot into the correct partion.

This way, Recalbox would be smaller and users could choose to install KODI, Raspbian etc.

You can then have KODI addons to support booting back into Recalbox. I made one here:
https://mega.nz/#!zdZCVSJC!L48HkCkT8DfPyjr-HgZFp2pnCTrz9q-RTQQDeEV0mG0
You simply set the boot partion in the KODI settings (defaults to 5), and it will reboot to that partion.

Or, they could simply reboot and choose Recalbox from NOOBS menu

You guys then don't have to worry about maintaining your KODI code. Just remove it.
The rebooting between OS's doesn't take that long.
You don't switch between gaming and media that often anyway.

Just seems logical too me to separate out the code. Leave the media center to the other guys.

Less work for you guys to maintain it, and user get's a "vanilla" KODI of their choice (OpenELEC or OSMC).

**Comment by matthuisman** [comment]
Here is a quick process

[code]
Recalbox boots
Search for other boot partions
If other partions:
     for each other partion:
           Get name of OS on partion eg. "OSMC" (from os.json - maybe get icon as well)
           Get boot partion number for OS eg. "5"
           Create new emulation station emulator entry "Operating Systems"
           Add rom entry "OSMC"

    User selects OS
    Run script:
         echo <BOOT_PARTION > /sys/module/bcm2709/parameters/reboot_part
         reboot
[/code]

**Comment by matthuisman** [comment]
<system>
<name>OS</name>
<fullname>Operating Systems</fullname>
<path>/oslist</path>
<extension>.os</extension>
<command>launchos.sh %ROM%</command>
</system>

/oslist folder  (generated on boot)
5.os
8.os

Then, could have metadata for these so they are shown nicely in the GUI

launchos.sh %ROM%

echo %ROM% /sys/module/bcm2709/parameters/reboot_part
reboot

(would need to strip .os from filename first)

**Comment by Sabcoll** [comment]
+1

it sucks that you cant install other stuff via apt-get.
no rpi-monitor, no hyperion for ambilight.

so please, give us the choice to start another OS from Recalbox!

**Comment by digitalLumberjack** [comment]
@matthuisman we'll think about that
@Sabcoll you can already start an other os with the noobs dual boot. https://github.com/digitalLumberjack/recalbox-os/wiki/Setup-a-dual-boot-raspbian-recalboxOS-%28EN%29
You should understand that all the features like lightweight, installspeed, online update etc are here because we chose to make a distribution from scratch, and not an other raspbian based distribution. Recalbox is not meant to be your desktop linux with apt get and internet browsers etc. Take it for what i meant it to be : a console firmware. So if you want apt-get, please just install a linux distribution, not a console firmware.

**Comment by Sabcoll** [comment]
True - and i really appreciate it!
I already opened another pull request on noobs for CEC support, so i can easily switch between the OS in noobs. I got OSMC and Recalbox on noobs running, but i have to plug in a mouse/keyboard and that looks so terrible on my TV-Board ... WAF and stuff ;-)

Still, it would be awesome to get Hyperion support in Recalbox, but that's another topic.

**Comment by matthuisman** [comment]
@Sabcoll 

I have instructions here to make a nice dual boot image:
https://www.raspberrypi.org/forums/viewtopic.php?p=815954#p815954

My setup

NOOBS boots into last selected OS (which is always OSMC)
I have the addon in NOOBS which reboots into RecalBox partion
  (This doesn't change the NOOBS remembered selection -YAH!)
Edited the Recalbox KODI launch script to boot into the OSMC partion

**Comment by Sabcoll** [comment]
Holy wack, i will try this later.
If it works - you'd deserve an programing oscar!

Looking forward to try!

**Comment by matthuisman** [comment]
You may also want to add
````
/recalbox/scripts/kodilauncher.sh
````
to /recalbox/scripts/rsync-update/always-exclude.cfg

to stop updating replacing your edited KODI launch file :)

**Comment by matthuisman** [comment]
I have started work on a start-up script to populate a new system with installed os's

````
import os
import json
import shutil

path = "/recalbox/scripts/os/settings"
if not os.path.exists(path):
	os.mkdir(path)

os.system("mount -o ro /dev/mmcblk0p3 %s" % path)

f = open(os.path.join(path, "installed_os.json"))
systems = json.loads(f.read())
f.close()

for system in systems:
	if not system['name'].startswith('recalboxOS'):
		name = system['name']
		icon_path = system['icon'].replace("/settings", path)
		description = system['description']
		partion = system['partitions'][0]
		
		#copy icon path etc

os.system("umount -l %s" % path)
````

**Comment by Sabcoll** [comment]
I didnt have time to test your previous script.
Shall i wait for your new start-up script? :-)

keep up the good work!

**Comment by matthuisman** [comment]
If your just using KODI, then use my first method :)
That's what I will keep using.

My new script is more "proof of concept"
The very first time Recalbox loads, it would run and generate a new "OS" emulator entry that allows booting into the other installed OS's.

**Comment by PreferLurking** [comment]
What is the point of moving kodi out of recalbox? What problems result from having kodi in recalbox?

I am not in favour of a dual-boot architecture.
I am a fan of recalbox specifically because it offers both emulators and media centre software without dual-booting.

**Comment by matthuisman** [comment]
What is your argument against dual booting?
It only takes another 10 seconds.

Separating the projects means Recalbox can focus more on what they do best and leave the media center side of things to the other guys.
It is also makes Recalbox smaller for the people who don't want KODI.

I know personally I like having OSMC or OpenELEC as my KODI. 
A lot bigger community if things go wrong etc.

**Comment by PreferLurking** [comment]
The pi is a computer (multi purpose) not a gaming console (single purpose). It can do both gaming emulation and media centre functions. Imagine that your phone had to be rebooted to switch applications. When you switch from your browser to your word processor, rebooting is not required. Rebooting is a kludge. I would like a single 'distro-like' approach to support both functions. There is much value in a platform that can successfully integrate them.
I have not been following recalbox's open issues closely, help me out.Again: other than you don't want it there, what problem does kodi's presence create?


     On Sunday, September 27, 2015 6:12 PM, matthuisman <notifications@github.com> wrote:
   

 What is your argument against dual booting?
It only takes another 10 seconds.Separating the projects means Recalbox can focus more on what they do best and leave the media center side of things to the other guys.
It is also makes Recalbox smaller for the people who don't want KODI.I know personally I like having OSMC or OpenELEC as my KODI. 
A lot bigger community if things go wrong etc.—

**Comment by matthuisman** [comment]
First of all, I wouldn't call Gaming Consoles single purpose. 
Many of them have media players, internet browsers etc.

Dual boot has many advantages.
For starters, it can use a different kernel.
Recalbox won't need to maintain a kernel that also works with KODI's features.

It's also about not re-inventing the wheel.
OSMC and OpenELEC are both mature-ish reliable KODI platforms with a good community.
Why have your own (with the headaches it can cause) when they do it better for the sake of 10 seconds.

If I was the developer on Recalbox, I would prefer to focus on emulation rather than KODI and also half the size of my system.

If you had a problem with KODI in Recalbox - where would you post the Issue?

**Comment by PreferLurking** [comment]
The manufacturers of gaming consoles also see the value of integrating multiple functions without the need to reboot.
If dual-boot had advantages, smartphones, PCs, tablets and gaming consoles would all employ it. Its unnecessary complexity. Two separate, specialized kernels is not an advantage. Its a disadvantage. Dual-booting is a bad idea even if it works flawlessly. I have used linux since the last century and have never needed to dual-boot.
Solving the 'right problem' is preferred over solving 'some other problem right'. Both functions under one kernel is the right problem to solve. Dual-booting for each function is another problem that wastes valuable time solving. I think working to separate the emulation and media centre functions is effort expended in the wrong direction. It seems that you have your dual booting working beside recalbox with kodi. That's great, but I do not see any reason for the rest of us to follow you while you head off in the wrong direction. Its much better to have both functions working correctly under the same kernel. I think your wishes are misguided for the community, but neither of us is the developer for recalbox, so ultimately, this decision will be made by someone else for us.
In my experience, due to the read-only squashfs filesystems in openELEC, its extremely difficult to add any functionality to. Also, the openELEC 'community' is at best unhelpful and at worst full of hostile wannabees. Adding emulation to openELEC is a non-starter.

OSMC seems to be using RetroPie to add into OSMC. I much prefer recalbox to RetroPie.
You speak of 'headaches' associated with kodi, but after being asked twice, you fail to name one.For the third time, are there any known technical issues associated with kodi's within recalbox?

     On Sunday, September 27, 2015 7:22 PM, matthuisman <notifications@github.com> wrote:
   

 First of all, I wouldn't call Gaming Consoles single purpose. 
Many of them have media players, internet browsers etc.Dual boot has many advantages.
For starters, it can use a different kernel.
Recalbox won't need to maintain a kernel that also works with KODI's features.It's also about not re-inventing the wheel.
OSMC and OpenELEC are both mature-ish reliable KODI platforms with a good community.
Why have your own (with the headaches it can cause) when they do it better for the sake of 10 seconds.If I was the developer on Recalbox, I would prefer to focus on emulation rather than KODI and also half the size of my system.If you had a problem with KODI in Recalbox - where would you post the Issue?—

**Comment by matthuisman** [comment]
Current headache is version of KODI in Recalbox is 14.
I have a new official 7" Raspberry Pi screen.
OSMC has just added support for this. So, I can easily run update in OSMC and it works.
I don't need to update Recalbox.
Should Recalbox release new updates just for KODI? That doesn't make sense.
Should I open an issue on Recalbox about KODI? 

KODI itself is adding emulation / games support eventually anyway..:
http://forum.kodi.tv/forumdisplay.php?fid=194

So, that will take care of people who want a media center / emulator set-up.

I therefore think Recalbox should focus on being an emulator set-up and have the option to boot into other OS's if need-be.

**Comment by PreferLurking** [comment]
I have been waiting for recalbox v3.3.0 to go 'gold' before trying it out. Does it still have kodi v14 in it?

**Comment by PreferLurking** [comment]
Do you have buildroot skills?

**Comment by Niehztog** [comment]
I have to agree with matthuisman, outsourcing the mediacenter to a dual-boot setup is the most pragmatic solution with most benefits regarding to updates and optimization, as both systems (gaming and media center) are currently not at all integerated into one user interface and/or backend.

Of course having one application with one userinterface for both would be preferable, but this currently is just not an option. As long as this state persists, dual-booting makes sense! Thank you matthuisman for your instructions, they work well here.

**Comment by matthuisman** [comment]
@Niehztog 

Thank you for the feedback.

I just think I would rather the Recalbox developers put all their time into the emulator side of things without having to worry about updating / fixing KODI & all it's dependencies.

KODI eventually will have it's own emulator side of things, so I think Recalbox should work towards being a great stand-alone emulator front end for people building arcade boxes etc who don't want a media center. If they remain a nice NOOBS package, then users can dual-boot to a media center if need be.

I could also see Raspberry Pi adding them into the official NOOBS package list, and thus greatly increasing their user base.

That would be my goal if I was them.
Make a very solid reliable emulator front-end and try to get into the NOOBS official OS list.

Then, add in the code to allow booting into the other detected NOOBS OS's (this is a good feature over the RetroPie competitor).

**Comment by Phone-IX** [comment]
Thank you so much! I was actually working on the same problem with dualbooting recalbox with osmc. I was searching all kind of wikis, blogs and forums for help. This is the best solution for this problem I could find!

**Comment by matthuisman** [comment]
@Phone-IX 

Glad to hear.
I have now made a small blog post how to make the nice dual-boot

http://matthuisman.blogspot.co.nz/2015/09/dual-boot-recalbox-and-osmc.html

**Comment by strydyr** [comment]
@matthuisman: I just registered to say "Thank You" for this.  

I appreciate everything that Recalbox is trying to do, but the Kodi updates are a significant issue for me.  Dual boot seems like the best answer from a number of angles.

**Comment by matthuisman** [comment]
In my latest images I have made to do this, the method I use is less "intrusive" than previous.
It allows users to use either the "internal" KODI or the "external" KODI

This is added to the top of kodilauncher.sh

[code]
systemsetting="python /usr/lib/python2.7/site-packages/configgen/settings/recalboxSettings.pyc"
kodiPartion="`$systemsetting -command load -key kodi.partion`"

if [ "$kodiPartion" != "" ];then
        echo $kodiPartion > /sys/module/bcm2709/parameters/reboot_part
        reboot
        exit 1
fi

(REST OF CODE)
[/code]

I then add a new setting to recalbox.conf

[code]
## set to boot partion number of KODI or remove to use internal KODI
kodi.partion=8
[/code]

Maybe this could be incorporated into the code to allow for both scenarios?

**Comment by nadenislamarre** [comment]
This is a good idea.
However, i suggest something really less intrusive :

- in the shutdown menu of es, add an entry if some fat16/32 exists on the
system to allow to boot on it
- in the option, make the X button configurable : launch kodi / start on
partition X, ...

Nicolas

2015-11-23 22:31 GMT+01:00 matthuisman <notifications@github.com>:

**Comment by yusf** [comment]
@matthuisman Here's another praise for enabling RB-OSMC dualboot. :beers:

**Comment by nadenislamarre** [comment]
won't do.
too complex.
kodi in recalbox is getting better and better. ask for missing kodi features in recalbox instead.

**Comment by matthuisman** [comment]
Fair enough.
My whole idea was to remove kodi completely to make it less complex.

Once retroplayer is merged into Kodi, then we will have a media center that does all the games as well. In kodi 17, controllers will work. Emulators etc most likely in KODi 18.

So, I  believe the aim should be towards stand alone arcade box situations who want a simple Emulatstaion front-end instead of KODI frontend. 

Removing it would make the build smaller and take up less developer time.

**Comment by nadenislamarre** [comment]
the joystick rewrite in kodi was already supposed to be done in v16... one
year later the work is at the very Bégining and my discussion on the kodi
irc showed me that the reflexion is still not mature. i will remove kodi
the day i found it will be necessary. not because it is told that in 2
years perhaps somethong will work in kodi ;-)
